### DISCUSSION TITLE
Improve typing of handlers

### DISCUSSION THREAD

**Discussion by marcuslimdw** [zauberzeug/nicegui#3797] [Ideas / Feature Requests]
Currently, handlers are typed permissively e.g. `on_change` is typed `Optional[Callable[..., Any]]`. This provides a poor DX because autocomplete and type checking don't work as well.

I propose the following:
1. Define a new type `Handler` (see below).
2. Use this type to annotate handler arguments - for example, `on_change` would now be annotated `Handler[ValueChangeEventArguments]`

```python
from typing import Any, Callable, Optional, TypeAlias, TypeVar, Union

EventT = TypeVar("EventT")
Handler: TypeAlias = Optional[Union[Callable[[EventT], Any], Callable[[], Any]]]
```

**Comment by falkoschindler** [comment]
Thank for bringing this up, @marcuslimdw!
I like the idea of improving type annotations for event handlers.

The only downside I see in using type aliases might be that they "hide" the type behind the alias "Handler[ClickEventArguments]":

<img width="661" alt="Screenshot 2024-09-29 at 16 56 01" src="https://github.com/user-attachments/assets/72353404-0f24-403c-9a20-e3ecd5de016b">

But I guess this is better than "Callable[..., Any]". And it is widely known that a "handler" is a kind of callable.

Would you like to create a pull request? 🙂

**Reply by falkoschindler** [reply]
PR #3811 has just been released.
